### ISSUE TITLE
Adapting this as a new resource for FAC 11

### ISSUE THREAD

**Issue by oliverjam** [dwyl/learn-tdd#90]
Hi @nelsonic!

@finnhodgkin and I are mentoring Testing Week so we're currently going through all the workshops we did.

We're going to be switching from QUnit to Tape (as per [this proposal](https://github.com/foundersandcoders/master-reference/issues/253)), so we need to re-work lots of the existing material.

We love this workshop and think it's a great resource as it is, so we wouldn't really want to change it. The vending machine change challenge fits TDD really well, so we'd love to keep using the concept.

We're wondering if you'd be happy with us creating a new workshop inspired by this one, but trimmed down and using Tape instead?

**Comment by nelsonic** [comment]
@oliverjam _all_ good questions, apologies for not making these clear sooner. 

The question of which testing library/framework comes up `before` and I'm _still_ not convinced that using Tape is _simpler_ than QUnit given that this tutorial is a _single_ `.html` file.

Our mission when writing a _beginner-friendly_ tutorial is to _assume_ as little as possible before people start, are you _aware_ if `tape` can be run _without_ Node.js?

To be _clear_ I think `tape` is _better_ for Node.js testing that's why the "next step" after learning TDD is https://github.com/dwyl/learn-tape but in order to _lower_ the "_barrier_" having the _single_ `.html` file for this tutorial has proved _very_ successful. 

With regards to creating a `new` workshop Is there a _specific_ advantage to creating a `new` workshop instead of re-using this one? Is there an area we _aren't_ already covering?
This tutorial has been used by F&C people since F&C5 and _hundreds_ of people in the community but we would _love_ to _improve_ it if you have time.

**Comment by oliverjam** [comment]
@nelsonic I think the distinction is that although this is a great workshop for introducing testing, we don't think it fits with the overall week.

To clarify: we decided that QUnit wasn't a great way to introduce testing, because lots of our cohort found it confusing to learn a methodology that isn't really used again after this week. It also encouraged some bad practices as it gives you access to the DOM (lots of people were writing code to check that their headings had the correct text for example).

QUnit also requires copying a large amount of boilerplate HTML to get it working, which I don't think is any less confusing than being told to run `npm install -D tape`.

>assume as little as possible before people start

I think this is a great goal for a general beginner tutorial, but when we control the entire learning environment I think we can be a bit more specific about what and how we teach people. Students will have to learn npm by week 4 anyway, and it's not like they suddenly become more capable at that point, so we don't think it's unreasonable to teach it at this point instead.

The advantage to creating a new workshop is that it would use Tape. I don't think there's anything missing from this workshop, it's just using a different technology to the rest of our planned week. We don't think this workshop should be changed to remove QUnit, which is why we were suggesting creating a totally new one with a similar coding challenge, but reworked to use Tape.

**Comment by nelsonic** [comment]
@oliverjam I'm _confused_. 
Are you suggesting that learning about testing/TDD should be _ignored_ until "**Node.js Week**"?

The _purpose_ of the "Learn TDD" tutorial was to teach testing as _early_ as _possible_ without requiring Node.js. 
"Learn TDD" has _one_ `index.html` file with **33 lines** of HTML 
which is arguably _not_ a "_large amount of boilerplate_", 
given that most people learning JS `should` already be familiar with HTML and loading a stylesheet or external script ... (_or is this an invalid assumption_...?)
Each line of the `index.html` is either explained in the readme or with an in-line comment.

We have _tested_ this tutorial on people at Codebar who _barely_ know any HTML and they have learned TDD in a couple of hours. 

Throw in Node.js `before` people are `.ready()` and they get "_overwhelmed_" _really fast_. 

> The _original_ idea behind teaching Testing (_which was an **afterthought** `before` F&C 5_) 
was to cover Testing as _early as possible_ 
(_in F&C 5 we covered it on the **second day** of **Week 1**! i.e. long `before` node.js_) 
We instilled good habits _early_ and that whole cohort has reaped the benefit.
(_there are a few who chose to work for companies that don't write any tests, 
but that's sadly beyond the control of F&C mentors; we did our best to give them good habits._) 

https://github.com/dwyl/learn-tape is _designed_ to follow on _seamlessly_ from "Learn TDD"... 
did you spot something _specific_ missing or unclear?

> To be clear: I don't think there should be a "testing week" or even a "testing day" ...
> I think testing should be learned from "Day 0" and **_every_ day** should be **_testing_ day**.
> Otherwise people rapidly descend into chaos/bad habits which are hard to break.

Everyone should have the ***10 Step Workflow*** "drummed" into them from the start:
+ Identify the need/requirement or pain point.
+ Write a user story (_as a GitHub Issue_) with acceptance criteria checklist.
+ Estimate how much time is required to build the story/feature.
+ Convert the user story into a test case (_or multiple test cases_)
+ Write code to pass the test case(s)
+ Commit code on a branch referencing the user story on GitHub.
+ Create Pull Request with feature referencing story
+ Submit PR to peers for review.
+ Reviewer comments, requests changes and ultimately merges into master.
+ Deploy (_using continuous deployment_)

Rinse. Repeat. Collect $200.

**Comment by iteles** [comment]
I'm with @nelsonic here, learning tape before you know anything about node will be pretty confusing.

You are of course more than welcome to include any dwyl workshops/tutorials as part of the course @oliverjam and if you feel they could be simplified once they have been tested, PRs are welcome!

**Comment by oliverjam** [comment]
@nelsonic Our cohort decided not to teach QUnit, and to teach Tape instead ([here is the proposal](https://github.com/foundersandcoders/master-reference/issues/253)). That's the only major change we want to make to Testing Week (Week 2 of the course).

The [current plan](https://github.com/foundersandcoders/master-reference/issues/299) is to go over the basics of npm at the start of the week, show the students how to set up Tape and then let them get on with writing tests. Why do you feel like learning about npm in Week 2 would be overwhelming, but not in Week 4?

If there's significant push back on using Tape in Week 2 then that's a conversation that needs to happen asap, because we've already finished our scheduled curriculum planning time.

**Comment by sohilpandya** [comment]
@oliverjam this proposal was not discussed at the meeting last week and there have been concerns raised by @des-des about extra dependencies in that issue too. https://github.com/foundersandcoders/master-reference/issues/253

It's not about learning npm in week 2 vs week 4. It's about trying to have something to build upon on a week by week basis. Otherwise, we would just introduce all the topics in the first couple of weeks as workshops and then let you go off and do whatever you wanted.

The **focus** of the week is testing and there are bound to be more questions about npm rather than testing if we introduce that in week 2. 

I would recommend you use still use qunit.

**Comment by des-des** [comment]
@iteles @nelsonic @sohilpandya @oliverjam 

I am not sure this issue is the place to challenge fac10's decision to use tape instead of qunit in week 2.

It seems to me oliver asked a pretty straight forward question. And ended up being drawn into a tangential discussion, rather than be able to focus his energy on achieving his goals.

This issue has been open for 20 days, https://github.com/foundersandcoders/master-reference/issues/253 and no one has objected.

**Comment by finnhodgkin** [comment]
@sohilpandya I disagree with the notion that teaching introductory npm in week 1 would disrupt the curriculum's week by week momentum. It's not as if we're advocating a Node core server in week 2, we're just suggesting we keep the same testing framework throughout to keep consistency. FAC10 had multiple SGCs with complaints about QUnit and Blanket.js; and it was also one of the most raised issues at our first curriculum planning session.

Plus, we actually currently **do** introduce npm in week 2. Blanket.js requires live server to work (as well as just not working with newer versions of QUnit).

**Comment by finnhodgkin** [comment]
Ee just noticed @des-des replied after you. Shall we move this conversation to the issue he mentioned?

**Comment by iteles** [comment]
Again, @oliverjam, you're more than welcome to use any dwyl tutorials and workshops to teach at Founders & Coders. Opening issues (and subsequent PRs) on simplifications, suggestions and feedback are most appreciated so we can continue to build on both the dwyl and Founders & Coders communities' needs 👍 

Let's move the discussion on whether or not QUnit should be part of the curriculum back to https://github.com/foundersandcoders/master-reference/issues/253.

**Comment by YvesMuyaBenda** [comment]
@nelsonic The following commentary is not really my thing to be involved in; but on a more general level, your comments on the purpose and context in which the tutorial was created in was _interesting_, and makes me wish all the tutorials had that kind of meta-data.

**Comment by nelsonic** [comment]
@oliverjam / @finnhodgkin, yes please move/keep any technology (_opinion-based_) discussions to where they can be debated (_by the **least experienced** but **loudest** voices_). 👍 
Thanks! ✨ 

Questions related to the _content_ of this tutorial should `continue` to be raised as issues here. ✅ 

# tl;dr

> I selected QUnit on the basis that it is "_neutral_" i.e. _unopinionated_. 
But as with _all_ decisions, there's _always_ going to be _someone_ who has a "_better_" idea.
(_which usually isn't "better", it's just someone's opinion based on incomplete information._)

I originally wrote this tutorial using `Mocha` which was the _early leader_ in Node.js Testing; 
Mocha is _still_ ***way more "popular"*** than Tape, Jest or Ava but that does not make it "_better_"!
After making the time to _understand_ Mocha in depth, I realised that it uses `Globals` for _everything_ and promptly stopped using it.

I have _nothing_ against people using Tape.
What I _object_ to is people _flagrantly_ copying dwyl's tutorials without _any_ attempt at attribution 
(_this has happened **several** times in F&C ... it's infuriating but no point "fighting" it._)

## Why Not _Jest_?

If we teach everyone to use [`Jest`](https://github.com/foundersandcoders/master-reference/issues/253#issuecomment-301588452) we'll be one step closer to getting people jobs at Facebook. 🙃 
That's _clearly_ what `some` people want because they don't _care_ about _selling out_ people's personal data to the highest bidder or selling elections to corporate interests.
I don't have _time_ to argue with those people anymore.
https://www.google.com/search?q=facebook+panorama+documentary&tbm=vid

While it _pains_ me that people "graduating" from F&C have virtually _nothing_ to distinguish themselves from all the others who have been _brainwashed_ into believing that the Facebook "stack" is somehow "_better_", I don't have _time_ to "_argue_" with anyone anymore; 
how do you _reason_ with a _zombie_?

![zombie](https://media.giphy.com/media/zJ61IgYwU9C92/giphy.gif)

I've _clearly stated_ my technology choices in https://github.com/dwyl/technology-stack
based on 7 years of using Node.js/JavaScript in Production at _several_ companies 
and realising that JavaScript is a actually _not_ the _best_ at _anything_ it does ... 
it's a _compromise_ I simply don't _have_ to make! 

_Unsurprisingly_ more people from _outside_ F&C have "_starred_" the repo than in
because F&C is has become an _echo-chamber_ for Facebook's marketing machine.

I intend to _fix_ this problem ***once-and-for-all*** as soon as I `finish` working on client projects 
and have time to dedicate to completely re-writing all the dwyl tutorials into a cohesive book with _extensive_ examples.

I _cannot_ even begin to describe how _delighted_ I am each time I write a test in Elixir and don't have to _think_ about "_which testing framework_"  I should use ... 
I'm _so_ glad that Facebook will _never_ get it's grubby/greedy paws on Elixir; because they cannot "own" it.

@des-des FYI: I have _totally_ given up on "_objecting_" to issues on Founders & Coders repositories because those discussions are not based on "_shoshin_" but rather on _fashion_ and the _opinions_ of people who have been _brainwashed_ by Facebook's marketing hype.

I find it _much_ easier to just `continue` getting my work done and if the _internet_ finds my output useful, that's great, but if not, I'm not the _slightest_ bit bothered. 

> Sadly, most people simply _don't care_ that Facebook enriching themselves 
by selling your personal data to corporate interests and greedy politicians. 😞 
Ignoring the fact that there is _objectively better_ tech than _anything_ in Facebook's "Stack", 
each time we _use_ something they make/own we are _directly_ supporting them by "_voting_" with our usage.
